File: leetcode/0010-regular-expression-matching/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
        dp[0][0] = true;

        for (int j = 1; j <= n; j++) {
            if (p[j - 1] == '*' && j >= 2) {
                dp[0][j] = dp[0][j - 2];
            }
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                char pc = p[j - 1];
                if (pc == '*') {
                    bool zeroOccurrence = dp[i][j - 2];
                    bool oneOrMore = false;
                    char prev = p[j - 2];
                    if (prev == '.' || prev == s[i - 1]) {
                        oneOrMore = dp[i - 1][j];
                    }
                    dp[i][j] = zeroOccurrence || oneOrMore;
                } else if (pc == '.' || pc == s[i - 1]) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = false;
                }
            }
        }

        return dp[m][n];
    }
};

File: leetcode/0010-regular-expression-matching/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[0][0] = True

        for j in range(1, n + 1):
            if p[j - 1] == "*" and j >= 2:
                dp[0][j] = dp[0][j - 2]

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                pc = p[j - 1]
                if pc == "*":
                    # p[j-2] is guaranteed to exist per the problem's constraints.
                    zero_occurrence = dp[i][j - 2]
                    one_or_more = False
                    prev = p[j - 2]
                    if prev == "." or prev == s[i - 1]:
                        one_or_more = dp[i - 1][j]
                    dp[i][j] = zero_occurrence or one_or_more
                elif pc == "." or pc == s[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = False

        return dp[m][n]
